File: 763.划分字母区间.cpp
/*
 * @lc app=leetcode.cn id=763 lang=cpp
 *
 * [763] 划分字母区间
 */

// @lc code=start
class Solution
{
public:
    vector<int> partitionLabels(string S)
    {
        int N = S.size();
        vector<int> ends(26, -1);
        for (int i = 0; i < N; ++i)
        {
            ends[S[i] - 'a'] = i;
        }
        vector<int> res;
        int i = 0;
        while (i < N)
        {
            int r = ends[S[i] - 'a'];
            for (int j = i + 1; j <= r; ++j)
            {
                r = max(r, ends[S[j] - 'a']);
            }
            res.push_back(r - i + 1);
            i = r + 1;
        }
        return res;
    }
};
// @lc code=end

File: 763.划分字母区间.java
import java.util.ArrayList;
import java.util.List;

/*
 * @lc app=leetcode.cn id=763 lang=java
 *
 * [763] 划分字母区间
 */

// @lc code=start
class Solution {
    public List<Integer> partitionLabels(String S) {
        int[] last = new int[26];
        int length = S.length();
        // record the last pos of per character
        for (int i = 0; i < length; i++) {
            last[S.charAt(i) - 'a'] = i;
        }
        List<Integer> partition = new ArrayList<>();
        int start = 0;
        int end = 0;
        for (int i = 0; i < length; i++) {
            // update the end pos by last pos
            end = Math.max(end, last[S.charAt(i) - 'a']);
            // when i equals end ,it means any character in the [start,end] don't appears in
            // the [end+1,S.length()-1]
            if (i == end) {
                partition.add(end - start + 1);
                start = end + 1;
            }
        }
        return partition;
    }
}
// @lc code=end
